indexed_gzip/zran_file_util.c

```diff
@@ -0,0 +1,210 @@
+    if ((len = PyBytes_Size(data)) == -1)
+        goto fail;
+    memmove(ptr, buf, (size_t) len);
+    Py_DECREF(data);
+    return (size_t) len / size;
+
+fail:
+    Py_XDECREF(data);
+    return 0;
+}
+
+/*
+ * Implements a method analogous to ftell that is performed on Python file-like objects.
+ */
+int64_t _ftell_python(PyObject *f) {
+    PyObject *data;
+    uint64_t result;
+    if ((data = PyObject_CallMethod(f, "tell", NULL)) == NULL)
+        goto fail;
+    if ((result = PyLong_AsUnsignedLong(data)) == (unsigned long)-1 && PyErr_Occurred())
+        goto fail;
+    Py_DECREF(data);
+    return result;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to fseek that is performed on Python file-like objects.
+ */
+int _fseek_python(PyObject *f, int64_t offset, int whence) {
+    PyObject *data;
+    if ((data = PyObject_CallMethod(f, "seek", "(l,i)", offset, whence)) == NULL)
+        goto fail;
+    Py_DECREF(data);
+    return 0;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to feof that is performed on Python file-like objects.
+ * This method requires the file size to be input, because there is no other way to tell if a Python file-like is at EOF without reading it.
+ */
+int _feof_python(PyObject *f, int64_t size) {
+    return _ftell_python(f) == size;
+}
+
+/*
+ * Implements a method analogous to ferror that is performed on Python file-like objects.
+ */
+int _ferror_python(PyObject *f) {
+    return PyErr_Occurred() ? 1 : 0;
+}
+
+/*
+ * Implements a method analogous to fflush that is performed on Python file-like objects.
+ */
+int _fflush_python(PyObject *f) {
+    PyObject *data;
+    if ((data = PyObject_CallMethod(f, "flush", NULL)) == NULL) goto fail;
+    Py_DECREF(data);
+    return 0;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to fwrite that is performed on Python file-like objects.
+ */
+size_t _fwrite_python(const void *ptr, size_t size, size_t nmemb, PyObject *f) {
+    PyObject *input;
+    PyObject *data = NULL;
+    unsigned long len;
+    if ((input = PyBytes_FromStringAndSize(ptr, size * nmemb)) == NULL)
+        goto fail;
+    if ((data = PyObject_CallMethod(f, "write", "(O)", input)) == NULL)
+        goto fail;
+    #if PY_MAJOR_VERSION >= 3
+    if ((len = PyLong_AsUnsignedLong(data)) == (unsigned long)-1 && PyErr_Occurred())
+        goto fail;
+    #else
+    // In Python 2, a file object's write() method does not return the number of
+    // bytes written, so let's just assume that everything has been written properly.
+    len = size * nmemb;
+    #endif
+    Py_DECREF(input);
+    Py_DECREF(data);
+    return (size_t) len / size;
+
+fail:
+    Py_XDECREF(input);
+    Py_XDECREF(data);
+    return 0;
+}
+
+/*
+ * Implements a method analogous to getc that is performed on Python file-like objects.
+ */
+int _getc_python(PyObject *f) {
+    char buf [1];
```

Comment:
Not that it matters much, but can't you declare `char buf;` and then `_fread_python(&buf,...`?

Reply:
Yeah, that's simpler. Changing it.